[SPARK-11865] [network] Avoid returning inactive client in TransportClientFactory.

vanzin opened the pull request:
There's a very narrow race here where it would be possible for the timeout handler
to close a channel after the client factory verified that the channel was still
active. This change makes sure the client is marked as being recently in use so
that the timeout handler does not close it until a new timeout cycle elapses.

Commit:
[SPARK-11865] [network] Avoid returning inactive client in TransportClientFactory.

There's a very narrow race here where it would be possible for the timeout handler
to close a channel after the client factory verified that the channel was still
active. This change makes sure the client is marked as being recently in use so
that the timeout handler does not close it until a new timeout cycle elapses.

SparkQA commented:
**[Test build #46377 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46377/consoleFull)** for PR 9853 at commit [`f836391`](https://github.com/apache/spark/commit/f83639155646684e17087c895150d7304b181ee0).

SparkQA commented:
**[Test build #46377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46377/consoleFull)** for PR 9853 at commit [`f836391`](https://github.com/apache/spark/commit/f83639155646684e17087c895150d7304b181ee0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46377/
Test FAILed.

vanzin commented:
Another mysterious failure where there are no failures... retest this please

SparkQA commented:
**[Test build #46431 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46431/consoleFull)** for PR 9853 at commit [`f836391`](https://github.com/apache/spark/commit/f83639155646684e17087c895150d7304b181ee0).

SparkQA commented:
**[Test build #46431 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46431/consoleFull)** for PR 9853 at commit [`f836391`](https://github.com/apache/spark/commit/f83639155646684e17087c895150d7304b181ee0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46431/
Test PASSed.

vanzin commented:
/cc @rxin @zsxwing ; we should probably get this into 1.6

zsxwing commented:
Since `ctx.close()` is asynchronous, this one doesn't fix the race totally. Right?

Commit:
Explicitly mark the client as timed out.

Since ctx.close() is asynchronous, this ensures that threads checking
for the client being alive get the right result.

vanzin commented:
@zsxwing true. Fixed.

zsxwing commented:
LGTM pending tests.

SparkQA commented:
**[Test build #46541 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46541/consoleFull)** for PR 9853 at commit [`83188ce`](https://github.com/apache/spark/commit/83188ce722c15ab10ebd764dd15ce4acb5d8714b).

SparkQA commented:
**[Test build #46541 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46541/consoleFull)** for PR 9853 at commit [`83188ce`](https://github.com/apache/spark/commit/83188ce722c15ab10ebd764dd15ce4acb5d8714b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46541/
Test PASSed.

vanzin commented:
Merging to master / 1.6.
